includes/class-amp-theme-support.php

```diff
@@ -1542,15 +1542,29 @@ public static function ensure_required_markup( DOMDocument $dom, $script_handles
 			$head->appendChild( $script );
 		}
 
-		// Ensure rel=canonical link.
-		$links         = [];
+		// Gather all links.
+		$links         = [
+			'preconnect' => [
+				// Include preconnect link for AMP CDN for browsers that don't support preload.
+				AMP_DOM_Utils::create_node(
```

Comment:
Only room for improvement here is that I'd like it if the page already had this link that it would be prevented from being duplicated.

Reply:
That would then come from a third-party plugin/theme instead? It would be difficult to detect, it might be easier to sanitize the duplicate away instead.

Reply:
Potentially. We actually can detect because we gather up a list of all the links right here. So we could just skip adding it if it already exists.

Aside: There is also a need to remove component scripts that do not exist among `$script_handles`, to remove scripts that are not relevant to the page (in case someone manually adds a script which turns out to not be relevant to the current page).

Reply:
In practice there should not be duplicates. After thinking about it more, it seems wasteful to detect for a duplicate `link[rel=preconnect]` for AMP, and it doesn't hurt to have it duplicated in the first place.